django/db/models/query.py

```diff
@@ -1098,6 +1098,13 @@ def ordered(self):
         else:
             return False
 
+    @property
+    def empty(self):
+        """
+        Return True if the QuerySet is empty (EmptyQuerySet).
+        """
+        return isinstance(self, EmptyQuerySet)
```

Comment:
this couples `QuerySet` to its subclass `EmptyQuerySet`, which is bad OOP, in fact impossible in most languages! Instead I think there it be a class-level attribute that's `False`, overridden on `EmptyQuerySet` to be `True`.

Also I'd suggest `is_empty` over `empty`, as `.empty()` sounds like it could be a method to me, as it works as a verb as well.

Reply:
but I found many usages like this, such as 
``` Python
    def union(self, *other_qs, all=False):
        # If the query is an EmptyQuerySet, combine all nonempty querysets.
        if isinstance(self, EmptyQuerySet):
            qs = [q for q in other_qs if not isinstance(q, EmptyQuerySet)]
            return qs[0]._combinator_query('union', *qs[1:], all=all) if qs else self
        return self._combinator_query('union', *other_qs, all=all)
```

Reply:
@adamchainz And I think the EmptyQuerySet is not a subclass of QuerySet.
```python
class InstanceCheckMeta(type):
    def __instancecheck__(self, instance):
        return isinstance(instance, QuerySet) and instance.query.is_empty()


class EmptyQuerySet(metaclass=InstanceCheckMeta):
    """
    Marker class to checking if a queryset is empty by .none():
        isinstance(qs.none(), EmptyQuerySet) -> True
    """

    def __init__(self, *args, **kwargs):
        raise TypeError("EmptyQuerySet can't be instantiated")

```

Reply:
you're right, my bad, it has been refactored I guess